### ISSUE TITLE
Use Wikipedia API instead of Wikipedia Mirror?

### ISSUE THREAD

**Issue by michaelsilver** [infolab-csail/WikipediaBase#31]
I think we can at least partially replace using the Wikipedia Mirror with the [Wikipedia API](https://www.mediawiki.org/wiki/API:Main_page), at least when fetching individual articles. Play around with their [API Sandbox](https://en.wikipedia.org/wiki/Special:ApiSandbox#action=query&titles=Main%20Page&prop=revisions&rvprop=content&format=jsonfm) to get a sense of the features. Basically, you send a HTTP GET query with the base address `https://en.wikipedia.org/w/api.php?` and then list a bunch of parameters. You can get the result back in both in JSON or XML, your choice, just change `format=xml` to `format=json`. We shouldn't get banned, per their explanation on [API Ediquette](https://www.mediawiki.org/wiki/API:Etiquette):
>Request limit

>There is no hard and fast limit on read requests, but we ask that you be considerate and try not to take a site down. Most sysadmins reserve the right to unceremoniously block you if you do endanger the stability of their site.

>If you make your requests in series rather than in parallel (i.e. wait for the one request to finish before sending a new request, such that you're never making more than one request at the same time), then you should definitely be fine. Also try to combine things into one request where you can (e.g. use multiple titles in a titles parameter instead of making a new request for each title. Or use "generator" if possible instead of making a request for each result from another request).

Ex: To get Barack Obama's rendered Wikipedia page:
```
https://en.wikipedia.org/w/api.php?action=query&titles=Barack%20Obama&prop=revisions&rvprop=content&rvparse=&format=xml
```
Ex2: To get Barack Obama's unrendered (wiki markup) page:
```
https://en.wikipedia.org/w/api.php?action=query&titles=Barack%20Obama&prop=revisions&rvprop=content&format=xml
```
NB: the page titles need to be escaped properly (e.g. spaces are `%20`, etc.). Here is the Ruby WikiBase function to escape titles:
```Ruby
  def self.uri_escape(str)
    str = URI.escape(str)

    # Encode those characters missed by URI.escape because it was
    # treating them as special characters in the context of URIs
    h = {
      "$" => "%24",
      "&" => "%26",
      "+" => "%2B",
      "," => "%2C",
      "/" => "%2F",
      ":" => "%3A",
      ";" => "%3B",
      "=" => "%3D",
      "?" => "%3F",
      "@" => "%40"
    }
    h.each do |k,v|
      str.gsub!(k,v)
    end
    str
  end
```

**Comment by fakedrake** [comment]
Yep, it could be implemented as a fetcher and it shouldn't be too hard either. I looked into it a bit last summer but I didn't find it to be significantly different to getting the real markup so I gave it a rest.

**Comment by alvaromorales** [comment]
I've been working on this (a `WikipediaAPIFetcher`). So far, the Wikipedia API greatly simplifies things:
* json response format, yay
* redirects are handled automatically (by Wikipedia)
* less issues with encodings. I'm using the `requests` library to parse the response and get back a unicode string. To output a string (e.g. to START, to stdout), I use `unidecode`.

This is exciting for at least two reasons:
* Live Wikipedia is fast! Scaling the wikipedia mirror and achieving good performance seems like a lot of work.
* Using the Wikipedia API could mean that we never have to "create / upgrade the backend" -- we always have the latest version.

I'll keep working on this and eventually file a PR.

**Comment by fakedrake** [comment]
That is exciting indeed. I am looking forward to it!

**Comment by alvaromorales** [comment]
Closed via #47
